alternator: rudimentary implementation of TTL expiration service

nyh opened the pull request:
In this patch series we add an implementation of an
expiration service to Alternator, which periodically scans the data in
the table, looking for expired items and deleting them.

We also continue to improve the TTL test suite to cover additional
corner cases discovered during the development of the code.
    
This implementation is good enough to make all existing tests but one,
plus a few new ones, pass, but is still a very partial and inefficient
implementation littered with FIXMEs throughout the code. Among other
things, this initial implementation doesn't do anything reasonable about pacing of
the scan or about multiple tables, it scans entire items instead of only the
needed parts, and because each shard "owns" a different subset of the
token ranges, if a node goes down, partitions which it "owns" will not
get expired.
    
The current tests cannot expose these problems, so we will need to develop
additional tests for them.
    
Because this implementation is very partial, the Alternator TTL continues
to remain "experimental", cannot be used without explicitly enabling this
experimental feature, and must not be used for any important deployment.

Refs #5060 but doesn't close the issue (let's not close it until we have a
reasonably complete implementation - not this partial one).

nyh commented:
Oops, I forgot to make the starting of the TTL scanning service (which in this implementation has horrible performance) conditional on the experimental TTL flag. I'll submit a new version soon.

psarna commented on test/alternator/test_ttl.py:
```diff
@@ -189,17 +190,28 @@ def test_update_ttl_errors(dynamodb):
+    if is_aws(dynamodb):
+        duration = 1200
+    else:
+        duration = 3
```
note: Python has a syntax that I personally find very nice and concise:
```python
duration = 1200 if is_aws(dynamodb) else 3
```
(Feel free to ignore this comment if you prefer the expanded version, I don't deeply mind either way)

nyh replied:
You're right. I'll use that syntax.

psarna commented on test/alternator/test_ttl.py:
```diff
@@ -294,6 +294,52 @@ def test_ttl_expiration(dynamodb):
+    if is_aws(dynamodb):
+        max_duration = 1200
+    else:
+        max_duration = 3
```
ditto about the `x if something else y` syntax

nyh replied:
Did this in the next version I'll send.

psarna commented on test/alternator/test_ttl.py:
```diff
@@ -402,6 +402,38 @@ def test_ttl_expiration_range(dynamodb):
+    if is_aws(dynamodb):
+        duration = 900
+    else:
+        duration = 3
```
ditto №3 - also, is there any particular reason why it's 900 here and 1200 in other test cases?

nyh replied:
The extreme slowness of the TTL taking effect on DynamoDB seems to depend on the schema of the table, so some tests were even slower than others. In one of the tests even 20 minutes consistently wasn't enough, so I changed it into an hour.  In any case, what really matters is the !is_aws case - nobody will run (or, will very rarely run) these tests on AWS again after the initial development.

By the way, note also that some tests have max_duration - there we can even list 10 hours and the test will finish when it finishes, probably faster - and some tests have "duration" where the test will last exactly that given time, so we want to make that time as low as possible.

Another by the way, running all the tests on Scylla isn't that "very slow" any more - it currently takes 18 seconds for all the ttl tests to finish, and I can possibly make it even faster. At some point I will need to consider to enable them by default, and not mark them "very slow" when running on Scylla.

nyh commented:
Pushed a new version that doesn't start the TTL expiration service (which currently has terrible performance) unless the experimental flag is on.

Note that I check this node's experimental flag - **not** the enabled-on-the-entire-cluster flag like I check in the UpdateTimeToLive request. This is deliberate: If there is some other node that doesn't support TTL it may change its mind at any moment and the entire cluster will need to start processing TTLs, and we will need to start the expiration service at that point - but don't have any mechanism to do it. Instead better just start the expiration service now, even though it won't have any real work to do until the TTL feature is finally enabled on the entire cluster.

psarna commented on alternator/serialization.cc:
```diff
@@ -272,6 +272,30 @@ big_decimal unwrap_number(const rjson::value& v, std::string_view diagnostic) {
     }
 }
 
+std::optional<big_decimal> unwrap_number(const rjson::value& v) {
```
It might be confusing to future users that one flavor of the function throws, and the other returns an optional, but they share a name. Perhaps this one could be `unwrap_number_noexcept`? Normally I wouldn't care, but lately my coroutinization patch broke /next because of non-obvious throw semantics in a function and it's still fresh in my memory (:

nyh replied:
I don't like the word "noexcept" because people would rightly expect that a function called `something_noexcept` would be, well, `noexcept`, but this function can't be because it calls `json::print` and `big_decimal::big_decimal` which can probably allocate and throw if out of memory.

Come to think of it, perhaps the strange function that "deserves" a strange name is the existing function, which takes a string and uses it to build an exception. I could call it `unwrap_number_throw` or `unwrap_number_checked` or something... Would that make sense?

By the way, now that I look at this again, I see something surprising - unwrap_number allows the element inside the JSON to be an actual number, instead of a string we expect it to be. Do you rememeber if this is deliberate? Does DynamoDB allow that in the input JSON? I'll write a test to check...

nyh replied:
Unrelated to this series but I'll throw the fix in here too - unwrap_number should not allow numbers in the JSON. Only strings.

nyh replied:
> Come to think of it, perhaps the strange function that "deserves" a strange name is the existing function, which takes a string and uses it to build an exception. I could call it `unwrap_number_throw` or `unwrap_number_checked` or something... Would that make sense?

Another reason I prefer this direction is that other functions we have, like unwrap_set and unwrap_list, also return some sort of empty value instead of throwing and don't take an error string as parameter. Unfortunately all these unwrap_* functions aren't consistent on how they mark an error (e.g., unwrap_list returns a null pointer, my new unwrap_number returns a null optional), but at least they are consistent about returning, not throwing, on parse error.

avikivity replied:
Could all it `try_unwrap_number()`.

nyh replied:
Ok, I'll use this try_unwrap_number() name.

psarna commented on alternator/ttl.cc:
```diff
@@ -113,4 +139,321 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+            schema_ptr s = cf.second->schema();
+            // Check if an expiration-time attribute is enabled for this
+            // table.
+            std::optional<std::string> attribute_name = find_tag(*s, TTL_TAG_KEY);
```
From caller's perspective, it looks like `find_tag` could just as well return an `optional<string_view>` and not own the string, which would save us a copy, since we never hold this `attribute_name` across a preemption point, and thus the string view can't get invalidated. On the other hand, these strings are most likely very small and perhaps it's better to obtain a copy here and make this function safe even if we want to use `attribute_name` across a preemption point in the future.

nyh replied:
If find_tag returns a string, it is just moved in this line 279, and not copied again. You're right that if find_tag returned a string_view and this code also saved a string_view, one copy would be saved. But it looks very scary - the pointer we're saving isn't to some object which the caller is holding (e.g., some document we're working on) - but rather to some transient list of tags. If a preemption point does happen - this pointer can get invalidated. So it seems like it's a bad idea to risk it - especially because this is not a frequent operation (this is just for checking once in a while on which attribute the ttl is enabled).

psarna commented on alternator/ttl.cc:
```diff
@@ -113,4 +139,321 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+        // FIXME: think about working on different tables in parallel.
+        // also need to notice when a new table is added, a table is
+        // deleted or when ttl is enabled or disabled for a table!
+        for (const auto& cf : _db.get_column_families()) {
```
I'm reasonably sure that in case when no tables specify a TTL it's still reasonably safe to go through this whole loop without a single preemption point, but we might want to revise that later when we strive for multi-tenancy. E.g. if there could be a million tables in the system, this tight loop might cause a potential reactor stall, because I think that it only has preemption points on the path that actually expires a TTL. Adding an arbitrary preemption point is easy with coroutines (`co_await coroutine::maybe_yield()`), but doing so exceedingly often is also bad, because it adds overhead. To sum up, I think that no action is necessary for now, I'm just leaving this note here for other reviewers to consider, maybe somebody has a different opinion on this subject.

Also, I see that you consider adding parallelism to this loop, in which case the problem would probably disappear on its own anyway.

nyh replied:
I'll add a call to `co_await coroutine::maybe_yield()` inside the loop, just in case.

psarna reviewed (commented):
Looks promising, I left some notes and nitpicks, but no blocking comments

avikivity commented on alternator/serialization.cc:
```diff
@@ -272,6 +272,30 @@ big_decimal unwrap_number(const rjson::value& v, std::string_view diagnostic) {
+        return std::nullopt;
+    }
+}
+
```
I think the other unwrap_number should be implemented in terms of this one.

avikivity commented on alternator/ttl.cc:
```diff
@@ -113,4 +139,321 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+// 3. If the expiration time is more than 5 years in the past, it is assumed
+//    to be malformed and ignored - and the item does not expire.
+static bool is_expired(gc_clock::time_point expiration_time, gc_clock::time_point now) {
+    constexpr gc_clock::duration five_years = std::chrono::seconds(5 * 365 * 24 * 3600);
```
You can write std::chrono::years(5);

nyh replied:
I actually tried that first - and it doesn't work. After spending 5 minutes on it I gave up and did other more important things.
My guess is that a "year" can't be translated to a fixed number of seconds because it depends on leap years and so on, so they didn't want to allow such conversion.

psarna replied:
You can go with `std::chrono::days` or `std::chrono::weeks` (: months can probably be tricky too, but we certainly don't have leap days. That problem seems weird though, according to https://en.cppreference.com/w/cpp/chrono/duration, std::chrono::years is a very specific ratio and does not consider leap years to be a problem. But you're right that this thing is probably not important enough to spend a considerable amount of time debugging it.

nyh replied:
I wonder if I just mispelled "years" and wrote "year" - which is a yet another class (https://en.cppreference.com/w/cpp/chrono/year) which is completely different and wouldn't work.
I'll try to use std::chrono::years, I think it's clearest and it doesn't matter if it covers another extra 5/4th of a day or something.

nyh replied:
So, std::chrono::years did work and I'll switch to it instead.
I am guessing the reason it didn't work previously was that I mistyped std::chrono::year, which also exists but is something else (and not convertable to seconds, obviously).

avikivity commented on alternator/ttl.cc:
```diff
@@ -113,4 +139,321 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+        for (const auto& cf : _db.get_column_families()) {
+            schema_ptr s = cf.second->schema();
+            // Check if an expiration-time attribute is enabled for this
+            // table.
```
It will be nicer to extract a single table's expiration into its own function.

nyh replied:
The reason I didn't do this yet is that I wasn't sure how I was going to organize this code in the long run. Will I have the individual tables being expired in parallel, or one expiration thread going over a piece of one table now, then a piece of a second table, and so on and so on.

Maybe I should have a function that scans a single table starting at a specific position - and then I can use this function either for scanning an entire table (as the code does today) or in the future for scanning a piece of a table starting in some position. I'll try to see how it works out.

nyh replied:
By the way, In the 12th patch I did introduce an inner function - not for an entire table but a specific range in a table. So I'll mark this resolved.

avikivity commented on alternator/ttl.cc:
```diff
@@ -113,4 +139,321 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+    // thread. Just runs run() in the background.
+    // FIXME: we need to have a way to stop the thread we started (abort_source)
+    if (_db.features().cluster_supports_alternator_ttl()) {
+        (void) run().handle_exception([] (std::exception_ptr ep) {
```
You should capture this future and await it in stop().

nyh replied:
Right. I left FIXMEs (as in the case for stopping) and need to start doing them either now, or after committing some initial version so I don't need to work on maintaining a large patch set and you don't need to review it over and over.

By the way, I just added paging to the code so at least that FIXME is done and now expiration works even on a larger table (still without any reasonable pacing).

avikivity commented on alternator/ttl.cc:
```diff
@@ -113,4 +139,321 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+            auto query_options = std::make_unique<cql3::query_options>(cl, std::vector<cql3::raw_value>{});
+            query_options = std::make_unique<cql3::query_options>(std::move(query_options), std::move(paging_state));
+            dht::partition_range_vector partition_ranges;
+            partition_ranges.push_back(dht::partition_range::make_open_ended_both_sides());
```
This should be limited to primary owned ranges, to make sure only one node at a time tries to expire a particular row, and to this shard's ranges.

nyh replied:
Yes, this is also in a FIXME above.

avikivity replied:
Was this FIXME fixed?

nyh replied:
Yes, in patch 12.. It looks like you want me to conflate the entire scanner implementation into one patch instead of introducing a simpler one (scanning everything) in this patch and then fixing it to scan only the data owned by each shard in patch 12. Ok then, I'll merge the two patches.

nyh replied:
The two patches were merged.

avikivity commented on main.cc:
```diff
@@ -1341,11 +1342,25 @@ int main(int ac, char** av) {
+                // start this outside the Alternator if().
+                if (cfg->check_experimental(db::experimental_features_t::ALTERNATOR_TTL)) {
+                    supervisor::notify("starting the expiration service");
+                    es.start(std::ref(db), std::ref(proxy)).get();
```
Should be run in the maintenance group to prevent dominating resources.

avikivity replied:
Or maybe we're already in the maintenance group, not sure.

avikivity commented:
Looks good, you listed many caveats, but I think it's pretty close to what we want.

nyh commented:
Pushed a new version. Addressed most review comments but not all, and a lot of things remain in FIXMEs.

The biggest FIXME fixed in this version is that the scan now uses paging to correctly scan the entire table - not just the first page. Of course, the code still scans the entire table on every shard in the system which is clearly not the efficient thing to do - but it passes the tests because they don't care that the same item gets scanned too much.

nyh commented:
Added another, 11th, patch, that implements proper support for clean shutdown of the expiration service (stopping the loops and waiting for them to stop). This implements a FIXME that I had in an earlier patch.

nyh commented:
Pushed a new version with just a 12th patch added for implementing splitting of the entire token range to all the different shards of the system, to avoid duplicate work and improve performance as each shard reads its own data (although, because of QUORUM, it also needs digests from a second node).

@psarna @avikivity @eliransin  I think we should try to converge on something we can commit now instead of continuing to develop this PR as a growing patch set. So please consider committing this version, or review it. Thanks.

avikivity commented on alternator/ttl.cc:
```diff
@@ -441,21 +447,35 @@ future<> expiration_service::run() {
             }
         }
         // FIXME: replace this silly 1-second sleep by something smarter.
```
What are we going to do about this?

nyh replied:
The plan is to have sort of "controller" which slows down the scan to the pace we need it to be, based on a parameter of expiration delay. If the user configures expiration delay of 48 hours, and we're scanning too fast and will finish in 2 hours, we should slow down the scan. On the other hand, for tests, I want the scan to run all the time, for the test to finish quickly.
But I wanted to fix this later.

avikivity commented on alternator/ttl.cc:
```diff
@@ -441,21 +447,35 @@ future<> expiration_service::run() {
-        co_await seastar::sleep(std::chrono::seconds(1));
+        try {
+            co_await seastar::sleep_abortable(std::chrono::seconds(1), _abort_source);
+        } catch(seastar::sleep_aborted&) {}
```
Shouldn't you co_return there?

nyh replied:
Good catch.
The weirid thing that it worked like this! Perhaps I tested the abort code in a slightly different version and then broke it. I'll test again.

nyh replied:
Oh, actually I think my code is correct.
If the sleep is aborted, all it does it to make it shorter. But then we go around the loop again, and notice the abort source was triggered, and co_return there. There is no need to do is specially in the catch.

psarna replied:
style nitpick: missing space between `catch` and `(`

avikivity commented on alternator/ttl.hh:
```diff
@@ -38,13 +39,25 @@ namespace alternator {
+    // running.
+    // shutdown() may be called even before start(), but may only be called once -
+    // calling it twice will result in an exception.
+    future<> shutdown();
```
Shouldn't it be called from stop()?

nyh replied:
Wow, I didn't realize that sharded_service has magic code in the case the class has a method called stop(), it calls it in each shard. Instead, I did this manually in main.cc:
```
                       es.invoke_on_all(&alternator::expiration_service::shutdown).get();
                       es.stop().get();
```
There is some benefit of renaming my shutdown() stop() and making the code one line shorter (no need for the explicit invoke_on_all). On the other hand, I have to be honest - I don't like the way that sharded_service has methods call start() and stop() and also call the object's *different* start() and stop() methods. It is very confusing.

Anyway, I'll change to stop() to conform with what the rest of the code does.

nyh replied:
Done in next version - renamed shutdown() to stop(), and then it gets automatically called from sharded_service::stop(). Added a comment to explain this, because it's not obvious.

avikivity commented on main.cc:
```diff
@@ -1343,6 +1343,7 @@ int main(int ac, char** av) {
 
             alternator::controller alternator_ctl(gossiper, proxy, mm, sys_dist_ks, cdc_generation_service, service_memory_limiter, *cfg);
             sharded<alternator::expiration_service> es;
+            std::any stop_expiration_service;
```
Don't we already have an abort source we can use, instead of creating a new one? Stopping the main loop shouldn't be done in many different ways.

nyh replied:
This stop_expiration_service isn't an abort_source, it's a deferred call for actually calling the stop.
The expiration service has a separate abort_source per shard, I don't understand how it can work differently. Yes, we should have a global per-shard abort source, but if we do that, we would have no control over destruction order of the services...
Don't the other services work exactly like I did it - each service has its own stop function, which uses its own abort source (or whatever) to stop its own work?

avikivity commented on alternator/ttl.cc:
```diff
@@ -261,6 +263,187 @@ static future<> expire_item(service::storage_proxy& proxy,
+    std::optional<dht::selective_token_range_sharder> _intersecter;
+public:
+    // FIXME: add a constructor to start iteration in the middle, to continue
+    // a scan after reboot.
```
I don't think it's really important. Restarts are rare enough so that we can continue from the beginning. Perhaps we should start from a random location to avoid favoring one area.

nyh replied:
The random location idea is an interesting solution, indeed much better than starting at min_token.
Because the scan period can be long (e.g., 48 hours), it is not inconcievable that we can have a restart in the middle of such long period and interfere with the expiration delay guarantees the user asked for. I agree it's not critical in most cases.

avikivity commented on alternator/ttl.cc:
```diff
@@ -261,6 +263,187 @@ static future<> expire_item(service::storage_proxy& proxy,
+        _intersecter.emplace(_s->get_sharder(), *_range_it, this_shard_id());
+        // _intersecter.next() can't be empty, because we can assume that
+        // every vnode must intersect with every shard (we don't have vnodes
+        // that aren't serviced by some of the shards). If we didn't make this
```
Why do you assume this?

avikivity replied:
Trivial counterexample: a vnode (7, 8].

nyh replied:
Oh :-(
I was assuming that vnodes are roughly of the same *large* size. I forgot they are completely random, so can be tiny.
Of course, it's bad when it's tiny, but I can't rule it out :-(
So I need to go back to the loop I was hoping to avoid. Oh, well.

nyh replied:
Will be fixed in the next version.

avikivity commented on alternator/ttl.cc:
```diff
@@ -261,6 +263,187 @@ static future<> expire_item(service::storage_proxy& proxy,
+    while (!p->is_exhausted()) {
+        if (abort_source.abort_requested()) {
+            co_return;
+        }
```
You can use abort_source::check() to make it clearer. Yes, it uses an exception, but in very exceptional circumstances.

nyh replied:
I did as you suggested, and didn't like the result. I am worried that eventually (it's not in the code yet), run() needs to ignore exceptions and retry again - we don't want any random failure like network error - to abort the TTL thread forever. Because we'll want to catch exceptions, it's inconvenient for the normal termination to also use exceptions. I think it's better to have a different path for the normal termination request.

avikivity commented on alternator/ttl.cc:
```diff
@@ -261,6 +263,187 @@ static future<> expire_item(service::storage_proxy& proxy,
+        // builtin limit of the page's size in bytes. Setting this limit to 1
+        // is useful for debugging the paging code with moderate-size data.
+        uint32_t limit = std::numeric_limits<uint32_t>::max();
+        std::unique_ptr<cql3::result_set> rs = co_await p->fetch_page(limit, gc_clock::now(), executor::default_timeout());
```
Is the default timeout appropriate?

We need an outer loop to restart in case we get an exception.

nyh replied:
Right. Adding FIXMEs for this.

avikivity commented on alternator/ttl.cc:
```diff
@@ -354,96 +537,13 @@ future<> expiration_service::run() {
-            dht::partition_range_vector partition_ranges;
-            partition_ranges.push_back(dht::partition_range::make_open_ended_both_sides());
-            auto p = service::pager::query_pagers::pager(s, selection, *query_state_ptr, *query_options, command, std::move(partition_ranges), nullptr);
-            while (!p->is_exhausted()) {
```
Please fold into the patch that introduced this.

nyh replied:
If I understand you correctly I assume you mean fold the patch which introduced scan-everything-on-each-shard with this patch that scans different ranges on each shard. Ok, will do.

nyh replied:
Combined the two patches.

nyh commented:
Pushed a new version, fixing many of the requests that Avi made. Most importantly, fixed one bug (we can't assume that every vnode will intersect with every shard), and merged three patches which implemented the scanning service incrementally (first scanning everything, then scanning only the parts owned by a shard) into one main patch - patch 7 - which implements the scanning service.

psarna commented:
thanks, I'll review the series today, might be in the evening though

psarna commented on alternator/ttl.cc:
```diff
@@ -113,4 +142,437 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+    if (n) {
+        return is_expired(*n, now);
+    }
+    return false;
```
nitpick: sounds like it can be folded to `return n && is_expired(*n, now)`

nyh replied:
Right :-) Will do it.

psarna commented on alternator/ttl.cc:
```diff
@@ -113,4 +142,437 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+        const auto& row_c = row[c];
+        if (!row_c) {
+            // This shouldn't happen - all key columns must have values.
+            // But if it ever happens, let's just *not* expire the item.
```
suggestion: if it shouldn't happen, let's log an error; if it somehow does happen, it will be easier to pinpoint later

nyh replied:
I'm a bit worried that if it does happen, it can happen a million times in a loop. I don't remember we have an existing mechanism for rate-limited logging, do we?

psarna replied:
If the message is going to be identical, syslog implementations are also free to collapse repetitions. In any case, when debugging, excessive logs are much better than no logs

psarna replied:
But yeah, I think that we only rate-limit some specific messages in our codebase and the support for this is open-coded, there's no generic mechanism for that

nyh replied:
I'm adding a FIXME for this. I don't know if logging is appropriate without good rate limiting - if anything is worse than suddenly TTL stopping to work it's suddenly the performance tanking because of excessive logging. Another option is to add a metric, but the whole metric thing is also left as a FIXME.

psarna commented on alternator/ttl.cc:
```diff
@@ -113,4 +142,437 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+            const auto& row_c = row[c];
+            if (!row_c) {
+                // This shouldn't happen - all key columns must have values.
+                // But if it ever happens, let's just *not* expire the item.
```
ditto - logging an error message could be helpful

nyh replied:
Also adding a FIXME here.

psarna commented on alternator/ttl.cc:
```diff
@@ -113,4 +142,437 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+        // the item to delete. FIXME: I feel there must be a much better
+        // way to look at results :-( Don't we have a scan API which will
+        // generate items with separately marked keys, instead of with this
+        // complicated result_set interface?
```
I think that the key columns appear in order, so you can use the first `n` for partition key, and then the next `m` for clustering key, with `n` and `m` being corresponding key sizes - assuming that you explicitly ask for all key values, but here you do. But I guess that your code is safer with regard to some corner cases I haven't thought of

nyh replied:
I'll take a look at if this idea can work - although this can also be a followup, because it's an optimization (and also listed as a FIXME).

I'm not happy at all about all the transformation that this code does - basically we read a key from the sstable as a contiguous serialized key, and then we open it up to that fancy "result set" API and then we need to wrap it back up to be a key. Everything would have been much more efficient if the key was just kept unparsed. But I think we can live with this inefficiency for now. It took me a long time to get this API to work, I don't want to start looking (just yet) for alternative APIs.

nyh commented:
I pushed a new version with a few minor changes requested in reviews and rebased.
A lot of FIXMEs, including things suggested by reviewers, are still in the code.
I think it's better at this point to commit this version (which is anyway protected by experimental flag) and work on it it incrementally, instead of adding more and more changes to this PR. It will be easier to review and to develop if we do it incrementally instead of fixing everything in the first pull request.

nyh commented:
I'm preparing a new version that fixes another FIXME, so please don't commit just yet (if you were thinking to).

nyh commented:
I pushed a new version, with the following improvements to the main patch ("alternator: initial implmentation of TTL expiration service"):

1. Created a function scan_table() for the main scanning loop of a single table. This was requested earlier in a review by @psarna, and also makes it easier to catch exceptions:
2. Fixed a FIXME where we need to need to catch exceptions during the scan. Not catching exceptions was obviously important (hence it was a FIXME, and @avikivity also noted it in a review), but even for an initial implementation it has practical implication: Even in simple tests, if a test deletes a table when it ends, and the scanning thread is scanning it in parallel, the scan would fail and abort the entire TTL-scanning thread. This bug was often visible when running the entire `test_ttl.py` test suite - often the TTL thread would abort in one of the tests, causing all the tests after it to fail. This bug is fixed in the new version.

nyh commented:
@avikivity @psarna I answered most of your concerns, and/or modified the code and/or added FIXMEs. Is there anything left to discuss?

I'd like us to please make a decision how two proceed with this pull-request:
1. One option is to commit it as-is (or are you waiting for any specific fixes?), and improvements will come incrementally later. This feature - and the background threads - only runs if an experimental feature is enabled.
2. The second option is for me to continue to make improvements and push better and better versions here, and not commit it until it has no obvious bugs or anti-features.

I'm very much in favor of option 1, but if you prefer option 2, please let me know so I can change my development process accordingly. Thanks.

CC @eliransin

psarna commented:
I prefer 1, I'll do another round of review now, and enqueue if I won't have any objections, ignoring the FIXME's for now.

psarna commented on alternator/ttl.cc:
```diff
@@ -113,4 +142,494 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+
+// Scan data in a list of token ranges in one table, looking for expired
+// items and deleting them.
+// Because of issue #9167, partition_ranges must have a single partition
```
btw - fix for #9167 landed on master; still, it's fine to address it in a follow-up to this series to avoid an endless loop of review

nyh replied:
Yes, I saw that, but 1. you're right about the endless loop, and 2. I'm not sure I trust this fix :-( I didn't see a unit test or any explanation of how it was tested, and now I no that nothing in the code uses multiple disconnected ranges, so I worry about being the first user. I'll think about this again later.

psarna commented on alternator/ttl.cc:
```diff
@@ -113,4 +142,494 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+                // a JSON serialized into bytes)
+                // FIXME: is it possible to find a specific member of a map
+                // without iterating through it like we do here and compare
+                // the key?
```
well, map is serialized as a vector of pairs, so you can use `find_if`, which is still iterating through the loop, but without open-coding it

psarna commented on alternator/ttl.cc:
```diff
@@ -113,4 +142,494 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+    // should be possible (and a must for issue #7751!).
+    lw_shared_ptr<service::pager::paging_state> paging_state = nullptr;
+    auto regular_columns = boost::copy_range<query::column_id_vector>(
+    s->regular_columns() | boost::adaptors::transformed([] (const column_definition& cdef) { return cdef.id; }));
```
this line misses indentation

psarna reviewed (approved):
Looks good. I'll sleep on it and enqueue it tomorrow in the morning -- and if somebody manages to veto this idea by then, I'll wait. Otherwise, let's proceed with option 1 and merge early, which also allows more people to take part in fixing the FIXME's

nyh commented:
Pushed a new version, rebased and with an additional patch at the end which fixes one FIXME:
In a FIXME I was worried that my big_decimal (arbitrary precision floating point) when checking for expiration was incorrect, and indeed I was able to verify in a test that it is incorrect for non-integer expiration times, and this new last patch adds this test - as well as a fix.
